x/stake/handler.go

```diff
@@ -178,62 +214,35 @@ func handleMsgDelegate(ctx sdk.Context, msg types.MsgDelegate, k keeper.Keeper)
 }
 
 func handleMsgBeginUnbonding(ctx sdk.Context, msg types.MsgBeginUnbonding, k keeper.Keeper) sdk.Result {
-	err := k.BeginUnbonding(ctx, msg.DelegatorAddr, msg.ValidatorAddr, msg.SharesAmount)
+	ubd, err := k.BeginUnbonding(ctx, msg.DelegatorAddr, msg.ValidatorAddr, msg.SharesAmount)
 	if err != nil {
 		return err.Result()
 	}
 
-	tags := sdk.NewTags(
-		tags.Action, tags.ActionBeginUnbonding,
-		tags.Delegator, []byte(msg.DelegatorAddr.String()),
-		tags.SrcValidator, []byte(msg.ValidatorAddr.String()),
-	)
-	return sdk.Result{Tags: tags}
-}
-
-func handleMsgCompleteUnbonding(ctx sdk.Context, msg types.MsgCompleteUnbonding, k keeper.Keeper) sdk.Result {
-
-	err := k.CompleteUnbonding(ctx, msg.DelegatorAddr, msg.ValidatorAddr)
-	if err != nil {
-		return err.Result()
-	}
+	finishTime := types.MsgCdc.MustMarshalBinary(ubd.MinTime)
 
 	tags := sdk.NewTags(
-		tags.Action, ActionCompleteUnbonding,
+		tags.Action, tags.ActionBeginUnbonding,
 		tags.Delegator, []byte(msg.DelegatorAddr.String()),
 		tags.SrcValidator, []byte(msg.ValidatorAddr.String()),
 	)
-
-	return sdk.Result{Tags: tags}
+	return sdk.Result{Data: finishTime, Tags: tags}
 }
 
 func handleMsgBeginRedelegate(ctx sdk.Context, msg types.MsgBeginRedelegate, k keeper.Keeper) sdk.Result {
-	err := k.BeginRedelegation(ctx, msg.DelegatorAddr, msg.ValidatorSrcAddr,
+	red, err := k.BeginRedelegation(ctx, msg.DelegatorAddr, msg.ValidatorSrcAddr,
 		msg.ValidatorDstAddr, msg.SharesAmount)
 	if err != nil {
 		return err.Result()
 	}
 
-	tags := sdk.NewTags(
-		tags.Action, tags.ActionBeginRedelegation,
-		tags.Delegator, []byte(msg.DelegatorAddr.String()),
-		tags.SrcValidator, []byte(msg.ValidatorSrcAddr.String()),
-		tags.DstValidator, []byte(msg.ValidatorDstAddr.String()),
-	)
-	return sdk.Result{Tags: tags}
-}
-
-func handleMsgCompleteRedelegate(ctx sdk.Context, msg types.MsgCompleteRedelegate, k keeper.Keeper) sdk.Result {
-	err := k.CompleteRedelegation(ctx, msg.DelegatorAddr, msg.ValidatorSrcAddr, msg.ValidatorDstAddr)
-	if err != nil {
-		return err.Result()
-	}
+	finishTime := types.MsgCdc.MustMarshalBinary(red.MinTime)
 
 	tags := sdk.NewTags(
-		tags.Action, tags.ActionCompleteRedelegation,
+		tags.Action, tags.ActionBeginRedelegation,
```

Comment:
Should we also tag the finish time?

Reply:
lol - I'm going to start using thumbs down more (you're great cwgoes though)

Reply:
https://github.com/cosmos/cosmos-sdk/pull/2405#discussion_r220820984

Reply:
https://github.com/cosmos/cosmos-sdk/pull/2405#discussion_r221189256